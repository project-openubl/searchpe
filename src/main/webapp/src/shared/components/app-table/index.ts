export { AppTable } from "./app-table";
