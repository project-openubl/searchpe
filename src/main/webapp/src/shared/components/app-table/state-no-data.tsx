import React from "react";
import {
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  Title,
} from "@patternfly/react-core";
import { AddCircleOIcon } from "@patternfly/react-icons";

export const StateNoData: React.FC = () => {
  return (
    <EmptyState variant={EmptyStateVariant.small}>
      <EmptyStateIcon icon={AddCircleOIcon} />
      <Title headingLevel="h2" size="lg">
        No data available
      </Title>
      <EmptyStateBody>
        Create a new resource to search using this table.
      </EmptyStateBody>
    </EmptyState>
  );
};
