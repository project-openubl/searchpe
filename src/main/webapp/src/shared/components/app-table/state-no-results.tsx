import React from "react";
import {
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  Title,
} from "@patternfly/react-core";
import { SearchIcon } from "@patternfly/react-icons";

export const StateNoResults: React.FC = () => {
  return (
    <EmptyState variant={EmptyStateVariant.small}>
      <EmptyStateIcon icon={SearchIcon} />
      <Title headingLevel="h2" size="lg">
        No results found
      </Title>
      <EmptyStateBody>
        No results match the filter criteria. Remove all filters or clear all
        filters to show results.
      </EmptyStateBody>
    </EmptyState>
  );
};
