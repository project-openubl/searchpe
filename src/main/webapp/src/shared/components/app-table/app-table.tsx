import React from "react";
import { Bullseye, Spinner, Skeleton } from "@patternfly/react-core";
import {
  Table,
  TableHeader,
  TableBody,
  ICell,
  IRow,
  IActions,
  ISortBy,
  IActionsResolver,
  IAreActionsDisabled,
  OnSort,
} from "@patternfly/react-table";

import { StateNoData } from "./state-no-data";
import { StateNoResults } from "./state-no-results";
import { StateError } from "./state-error";

export interface AppTableProps {
  columns: ICell[];
  rows: IRow[];
  actions?: IActions;
  actionResolver?: IActionsResolver;
  areActionsDisabled?: IAreActionsDisabled;

  isLoading: boolean;
  loadingVariant?: "skeleton" | "spinner" | "none";
  fetchError?: any;

  sortBy?: ISortBy;
  onSort?: OnSort;

  filtersApplied: boolean;
  noDataState?: any;
  noSearchResultsState?: any;
  errorState?: any;
}

const ARIA_LABEL = "App table";

export const AppTable: React.FC<AppTableProps> = ({
  columns,
  rows,
  actions,
  actionResolver,
  areActionsDisabled,
  isLoading,
  fetchError,
  loadingVariant = "skeleton",
  sortBy,
  onSort,
  filtersApplied,
  noDataState,
  noSearchResultsState,
  errorState,
}) => {
  if (isLoading && loadingVariant !== "none") {
    let rows: IRow[] = [];
    if (loadingVariant === "skeleton") {
      rows = [...Array(10)].map(() => {
        return {
          cells: [...Array(columns.length)].map(() => ({
            title: <Skeleton />,
          })),
        };
      });
    } else if (loadingVariant === "spinner") {
      rows = [
        {
          heightAuto: true,
          cells: [
            {
              props: { colSpan: 8 },
              title: (
                <Bullseye>
                  <Spinner size="xl" />
                </Bullseye>
              ),
            },
          ],
        },
      ];
    } else {
      throw new Error("Can not determine the loading state of table");
    }

    return (
      <Table aria-label={ARIA_LABEL} cells={columns} rows={rows}>
        <TableHeader />
        <TableBody />
      </Table>
    );
  }

  if (fetchError) {
    return (
      <>
        <Table aria-label={ARIA_LABEL} cells={columns} rows={[]}>
          <TableHeader />
          <TableBody />
        </Table>
        {errorState ? errorState : <StateError />}
      </>
    );
  }

  if (rows.length === 0) {
    return filtersApplied ? (
      <>
        <Table aria-label={ARIA_LABEL} cells={columns} rows={[]}>
          <TableHeader />
          <TableBody />
        </Table>
        {noSearchResultsState ? noSearchResultsState : <StateNoResults />}
      </>
    ) : (
      <>
        <Table aria-label={ARIA_LABEL} cells={columns} rows={[]}>
          <TableHeader />
          <TableBody />
        </Table>
        {noDataState ? noDataState : <StateNoData />}
      </>
    );
  }

  return (
    <Table
      aria-label={ARIA_LABEL}
      cells={columns}
      rows={rows}
      actions={actions}
      actionResolver={actionResolver}
      areActionsDisabled={areActionsDisabled}
      sortBy={sortBy}
      onSort={onSort}
    >
      <TableHeader />
      <TableBody />
    </Table>
  );
};
