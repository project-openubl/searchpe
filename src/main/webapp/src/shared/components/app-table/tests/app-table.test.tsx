import React from "react";
import { mount, shallow } from "enzyme";
import { Label, Skeleton, Spinner } from "@patternfly/react-core";
import {
  ICell,
  IRow,
  IActions,
  SortColumn,
  sortable,
} from "@patternfly/react-table";

import { AppTable } from "../app-table";
import { StateNoData } from "../state-no-data";
import { StateNoResults } from "../state-no-results";
import { StateError } from "../state-error";

describe("AppTable", () => {
  const columns: ICell[] = [
    { title: "Col1", transforms: [sortable] },
    { title: "Col2" },
    { title: "Col3" },
  ];
  const rows: IRow[] = [...Array(15)].map((_, rowIndex) => {
    return {
      cells: [...Array(columns.length)].map((_, colIndex) => ({
        title: <Label>${`${rowIndex},${colIndex}`}</Label>,
      })),
    };
  });
  const actions: IActions = [
    {
      title: "Action1",
      onClick: jest.fn,
    },
    {
      title: "Action2",
      onClick: jest.fn,
    },
  ];

  it("Renders without crashing", () => {
    const wrapper = shallow(
      <AppTable
        columns={columns}
        rows={rows}
        isLoading={false}
        filtersApplied={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Renders error", () => {
    const wrapper = mount(
      <AppTable
        columns={columns}
        rows={rows}
        isLoading={false}
        fetchError={"Any error"}
        filtersApplied={false}
      />
    );
    expect(wrapper.find(StateError).length).toEqual(1);
  });

  it("Renders loading with skeleton", () => {
    const wrapper = mount(
      <AppTable
        columns={columns}
        rows={rows}
        isLoading={true}
        loadingVariant="skeleton"
        filtersApplied={false}
      />
    );
    expect(wrapper.find(Skeleton).length).toBeGreaterThan(1);
  });

  it("Renders loading with spinner", () => {
    const wrapper = mount(
      <AppTable
        columns={columns}
        rows={rows}
        isLoading={true}
        loadingVariant="spinner"
        filtersApplied={false}
      />
    );
    expect(wrapper.find(Spinner).length).toBe(1);
  });

  it("Renders empty table without aplying filters", () => {
    const wrapper = mount(
      <AppTable
        columns={columns}
        rows={[]}
        isLoading={false}
        filtersApplied={false}
      />
    );
    expect(wrapper.find(StateNoData).length).toEqual(1);
  });

  it("Renders empty table after applying filters", () => {
    const wrapper = mount(
      <AppTable
        columns={columns}
        rows={[]}
        isLoading={false}
        filtersApplied={true}
      />
    );
    expect(wrapper.find(StateNoResults).length).toEqual(1);
  });

  it("Render rows with static actions", () => {
    const wrapper = mount(
      <AppTable
        columns={columns}
        rows={rows}
        actions={actions}
        isLoading={false}
        filtersApplied={false}
      />
    );
    expect(wrapper.find(Label).length).toEqual(45); // 3 columns * 15 rows
  });

  it("Renders on sort", () => {
    const onSortMock = jest.fn();

    const wrapper = mount(
      <AppTable
        columns={columns}
        rows={rows}
        isLoading={false}
        onSort={onSortMock}
        filtersApplied={false}
      />
    );

    wrapper.find(SortColumn).simulate("click");
    expect(onSortMock.mock.calls.length).toEqual(1);
  });
});
