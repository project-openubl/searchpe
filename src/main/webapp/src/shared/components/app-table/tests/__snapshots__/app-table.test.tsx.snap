// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTable Renders without crashing 1`] = `
<Table
  aria-label="App table"
  borders={true}
  canSelectAll={true}
  canSortFavorites={true}
  cells={
    Array [
      Object {
        "title": "Col1",
        "transforms": Array [
          [Function],
        ],
      },
      Object {
        "title": "Col2",
      },
      Object {
        "title": "Col3",
      },
    ]
  }
  className=""
  contentId="expanded-content"
  dropdownDirection="down"
  dropdownPosition="right"
  expandId="expandable-toggle"
  gridBreakPoint="grid-md"
  isStickyHeader={false}
  ouiaSafe={true}
  role="grid"
  rowLabeledBy="simple-node"
  rows={
    Array [
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              0,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              0,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              0,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              1,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              1,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              1,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              2,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              2,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              2,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              3,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              3,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              3,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              4,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              4,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              4,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              5,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              5,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              5,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              6,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              6,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              6,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              7,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              7,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              7,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              8,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              8,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              8,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              9,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              9,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              9,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              10,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              10,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              10,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              11,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              11,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              11,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              12,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              12,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              12,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              13,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              13,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              13,2
            </Label>,
          },
        ],
      },
      Object {
        "cells": Array [
          Object {
            "title": <Label>
              $
              14,0
            </Label>,
          },
          Object {
            "title": <Label>
              $
              14,1
            </Label>,
          },
          Object {
            "title": <Label>
              $
              14,2
            </Label>,
          },
        ],
      },
    ]
  }
  selectVariant="checkbox"
  variant={null}
>
  <TableHeader />
  <Component />
</Table>
`;
